### ISSUE TITLE
Neo4j Compatibility: py2neo Invalid transaction

### ISSUE THREAD

**Issue by Wenzel** [technige/py2neo#804]
Hi,

I have been using your library since a long time with Neo4j **3.4**, (py2neo **v4.3.0**, the latest).

But when I try to upgrade my Neo4j DB to **4.0.1**, my application immediately throws invalid transaction erros from `py2neo`:

~~~Python
  File "/home/xxx/hooks/xxx.py", line 476, in process_obj
    self.tx.create(obj)
  File "/home/xxx/venv/lib/python3.8/site-packages/py2neo/database.py", line 906, in create
    create(self)
  File "/home/xxx/venv/lib/python3.8/site-packages/py2neo/ogm.py", line 419, in __db_create__
    self.__db_merge__(tx)
  File "/home/xxx/venv/lib/python3.8/site-packages/py2neo/ogm.py", line 440, in __db_merge__
    tx.create(node)
  File "/home/xxx/venv/lib/python3.8/site-packages/py2neo/database.py", line 906, in create
    create(self)
  File "/home/xxx/venv/lib/python3.8/site-packages/py2neo/data.py", line 609, in __db_create__
    create_subgraph(tx, self)
  File "/home/xxx/venv/lib/python3.8/site-packages/py2neo/internal/operations.py", line 136, in create_subgraph
    for i, identity in enumerate(identities):
  File "/home/xxx/venv/lib/python3.8/site-packages/py2neo/internal/operations.py", line 87, in _create_nodes
    for record in tx.run(cypher, x=data):
  File "/home/xxx/venv/lib/python3.8/site-packages/py2neo/database.py", line 823, in run
    return Cursor(self.connector.run(statement=cypher,
  File "/home/xxx/venv/lib/python3.8/site-packages/py2neo/internal/connectors.py", line 294, in run
    return self._run_in_tx(statement, parameters, tx, graph, keys, entities)
  File "/home/xxx/venv/lib/python3.8/site-packages/py2neo/internal/connectors.py", line 259, in _run_in_tx
    self._assert_valid_tx(tx)
  File "/home/xxx/venv/lib/python3.8/site-packages/py2neo/internal/connectors.py", line 196, in _assert_valid_tx
    raise TransactionError("Invalid transaction")
py2neo.database.TransactionError: Invalid transaction
~~~

:arrow_right:  Is there a breaking change in transactions between Neo4j `3.4` and `4.0.1` ?

If that true, since py2neo is **compatible** from Neo4j 3.2 to 4.0, it should handle this (possible) corner case internally.

@technige what do you think ?
Currently I can't upgrade my Neo4j database because of this py2neo issue.

Thank you !

**Comment by dcsan** [comment]
have a look in the issues, there are quite a few related topics here. I ended up going back 3.x also because of this

**Comment by technige** [comment]
@Wenzel Have you been able to try with the latest v5 beta code? If not, can you. This should work with 4.0. If this still fails, could you share a snippet of code to help recreate the issue?

**Comment by Wenzel** [comment]
@technige thanks for your reply.

By trying to replicate this issue in a code snippet, I found out that the first transaction creation fails with an Exception that I thought was a simple warning, and simply displaying on the stdout:
~~~Python
Traceback (most recent call last):
  File "<ipython-input-2-d6070ded6757>", line 33, in <module>
    tx.create(sevenzip_inode)
  File "/home/wenzel/Projets/oswatcher/venv/lib/python3.8/site-packages/py2neo/database.py", line 906, in create
    create(self)
  File "/home/wenzel/Projets/oswatcher/venv/lib/python3.8/site-packages/py2neo/ogm.py", line 419, in __db_create__
    self.__db_merge__(tx)
  File "/home/wenzel/Projets/oswatcher/venv/lib/python3.8/site-packages/py2neo/ogm.py", line 444, in __db_merge__
    related_objects.__db_push__(tx)
  File "/home/wenzel/Projets/oswatcher/venv/lib/python3.8/site-packages/py2neo/ogm.py", line 278, in __db_push__
    tx.run("MATCH %s WHERE id(a) = {x} AND NOT id(b) IN {y} DELETE _" % self.__relationship_pattern,
  File "/home/wenzel/Projets/oswatcher/venv/lib/python3.8/site-packages/py2neo/database.py", line 823, in run
    return Cursor(self.connector.run(statement=cypher,
  File "/home/wenzel/Projets/oswatcher/venv/lib/python3.8/site-packages/py2neo/internal/connectors.py", line 294, in run
    return self._run_in_tx(statement, parameters, tx, graph, keys, entities)
  File "/home/wenzel/Projets/oswatcher/venv/lib/python3.8/site-packages/py2neo/internal/connectors.py", line 282, in _run_in_tx
    result.keys()   # force receipt of RUN summary, to detect any errors
  File "/home/wenzel/Projets/oswatcher/venv/lib/python3.8/site-packages/py2neo/internal/hydration/__init__.py", line 77, in keys
    self._on_more()
  File "/home/wenzel/Projets/oswatcher/venv/lib/python3.8/site-packages/py2neo/internal/connectors.py", line 262, in fetch
    tx.fetch()
  File "/home/wenzel/Projets/oswatcher/venv/lib/python3.8/site-packages/neobolt/direct.py", line 419, in fetch
    return self._fetch()
  File "/home/wenzel/Projets/oswatcher/venv/lib/python3.8/site-packages/neobolt/direct.py", line 461, in _fetch
    response.on_failure(summary_metadata or {})
  File "/home/wenzel/Projets/oswatcher/venv/lib/python3.8/site-packages/neobolt/direct.py", line 751, in on_failure
    handler(metadata)
  File "/home/wenzel/Projets/oswatcher/venv/lib/python3.8/site-packages/py2neo/internal/connectors.py", line 267, in fail
    self._fail(metadata)
  File "/home/wenzel/Projets/oswatcher/venv/lib/python3.8/site-packages/py2neo/internal/connectors.py", line 288, in _fail
    raise GraphError.hydrate(metadata)
py2neo.database.ClientError: SyntaxError: The old parameter syntax `{param}` is no longer supported. Please use `$param` instead (line 1, column 44 (offset: 43))
"MATCH (a)-[_:HAS_CHILD]->(b) WHERE id(a) = {x} AND NOT id(b) IN {y} DELETE _"
~~~

The message `The old parameter syntax {param} is no longer supported. Please use $param instead` is similar to this issue closed in April: https://github.com/technige/py2neo/issues/801.

So `py2neo 4.3.0` **is not compatible** with `Neo4j 4.0` ?? :thinking: 

I think this corrupts the internal state of `py2neo` and prevents me from doing any further `tx.create(node)`, leading to the `Invalid transaction` every time.

If you want to repro, you can use the following snippet:

~~~Python
import logging

from py2neo import Graph
from py2neo.ogm import GraphObject, RelatedTo, Property


class GraphInode(GraphObject):
    # properties
    path = Property()
    # relationships
    children = RelatedTo("GraphInode", "HAS_CHILD")

    def __init__(self, path):
        super().__init__()
        self.path = path


graph = Graph(password='admin')
tx = graph.begin()
sevenz_chm = GraphInode('/Program Files/7zip/7-zip.chm')
tx.create(sevenz_chm)
sevenz_dll = GraphInode('/Program Files/7zip/7-zip.dll')
tx.create(sevenz_dll)
sevenz_32dll = GraphInode('/Program Files/7zip/7-zip32.dll')
tx.create(sevenz_32dll)
sevenzip_inode = GraphInode('/Program Files/7zip')


sevenzip_inode.children.add(sevenz_chm)
sevenzip_inode.children.add(sevenz_dll)
sevenzip_inode.children.add(sevenz_32dll)
try:
    tx.create(sevenzip_inode)
except Exception as e:
    logging.exception('First transaction failed')

program_files = GraphInode('/Program Files')
program_files.children.add(sevenzip_inode)
tx.create(program_files)
~~~

**Comment by Wenzel** [comment]
Also, testing with the latest beta available, `5.0b1`, I'm facing the following exception:

~~~Python
TypeError                                 Traceback (most recent call last)
<ipython-input-11-c325f98c8ade> in <module>
     10 
     11 graph = Graph(password='admin')
---> 12 person = Person.match(graph).where(f"_.name = 'Tom Hanks'").first()

~/Projets/oswatcher/venv/lib/python3.8/site-packages/py2neo/ogm.py in match(cls, graph, primary_value)
    396         :rtype: :class:`.GraphObjectMatch`
    397         """
--> 398         return GraphObjectMatcher(cls, graph).match(primary_value)
    399 
    400     def __repr__(self):

~/Projets/oswatcher/venv/lib/python3.8/site-packages/py2neo/ogm.py in match(self, primary_value)
    503         cls = self._object_class
    504         if cls.__primarykey__ == "__id__":
--> 505             match = NodeMatcher.match(self, cls.__primarylabel__).where("id(_) = %d" % primary_value)
    506         elif primary_value is None:
    507             match = NodeMatcher.match(self, cls.__primarylabel__)

TypeError: %d format: a number is required, not NoneType
~~~

Raised by the following code snippet:
~~~Python
from py2neo import Graph
from py2neo.ogm import GraphObject, Property


class Person(GraphObject):
    name = Property()


graph = Graph(password='admin')
person = Person.match(graph).where(f"_.name = 'Tom Hanks'").first()
~~~

Without a defined `__primarykey__` attribute, the code raised this TypeError exception, because the default primarykey is set to `None`.

So, should I absolutely define a `primarykey` to use `py2neo v5` ?

Thanks !

**Comment by Wenzel** [comment]
@technige Actually I found the issue for Neo4j 4.0 compatibility.

Using tag `py2neo-4.3.0`:
py2neo's code is using the **old syntax**, as the error messsage says:
https://github.com/technige/py2neo/blob/py2neo-4.3.0/py2neo/ogm.py#L278
~~~Python
        tx.run("MATCH %s WHERE id(a) = {x} AND NOT id(b) IN {y} DELETE _" % self.__relationship_pattern,
               x=subject_id, y=[obj.__node__.identity for obj, _ in related_objects])
~~~

Should be
~~~Python
        params = {
            'x': subject_id,
            'y': [obj.__node__.identity for obj, _ in related_objects],
        }
        tx.run("MATCH %s WHERE id(a) = $x AND NOT id(b) IN $y DELETE _" % self.__relationship_pattern,
               parameters=params)
~~~

And all of other places in the code where you use the old parameter syntax needs to be updated:
- [ogm.py:__db_push__](https://github.com/technige/py2neo/blob/py2neo-4.3.0/py2neo/ogm.py#L282)
- [matching.py:_query_and_parameters](https://github.com/technige/py2neo/blob/py2neo-4.3.0/py2neo/matching.py#L314)
- [operations.py:pull_subgraph()](https://github.com/technige/py2neo/blob/py2neo-4.3.0/py2neo/internal/operations.py#L240)

etc.

Can you please fix this compatibility issue with Neo4j 4.0 on `v4`  ?

Related issues: https://github.com/technige/py2neo/issues/801 and https://github.com/technige/py2neo/issues/791
Thanks.

**Comment by technige** [comment]
I will only be fixing this in v5. For Neo4j 4.x compatibility, please use this version.
